### PR TITLE
Disable pattern matching in val syntax in sbt 0.13.7.

### LINKED ISSUE #1661: reflective compilation has failed: '=' expected but eof found in 0.13.7-M3
It appears that defining multiple projects in a single line `lazy val a, b = project` is broken in 0.13.7-M3. It works fine in 0.13.6 (and I'm sure worked fine in 0.13.7-M3 today, but just out of the sudden stopped).

The messages are:

```
➜  jacoco4sbt-root-no-sources  xsbt
JAVA_HOME=/Library/Java/JavaVirtualMachines/java8/Contents/Home
SBT_OPTS= -Xms512M -Xmx1536M -Xss1M -XX:+CMSClassUnloadingEnabled -Dfile.encoding=UTF-8
[info] Loading global plugins from /Users/jacek/.sbt/0.13/plugins
[info] Updating {file:/Users/jacek/.sbt/0.13/plugins/}global-plugins...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[info] Loading project definition from /Users/jacek/sandbox/jacoco4sbt-root-no-sources/project
[info] Updating {file:/Users/jacek/sandbox/jacoco4sbt-root-no-sources/project/}jacoco4sbt-root-no-sources-build...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[error] [/Users/jacek/sandbox/jacoco4sbt-root-no-sources/build.sbt]:2: reflective compilation has failed:
[error]
[error] '=' expected but eof found.
Project loading failed: (r)etry, (q)uit, (l)ast, or (i)gnore?
```

When I change `project/build.properties` to `sbt.version=0.13.6` it works fine.

### PR DESCRIPTION
**Pull request by jsuereth**
Fixes #1661 or at least works around it for now.

Review by @eed3si9n cc> @jaceklaskowski

### CODE REVIEW

File: main/src/main/scala/sbt/internals/parser/SbtParser.scala [hunk 1]
```diff
@@ -92,6 +92,21 @@ private[sbt] case class SbtParser(file: File, lines: Seq[String]) extends Parsed
         Seq(t)
     }
 
+    // Check No val (a,b) = foo *or* val a,b = foo as these are problematic to range positions and the WHOLE architecture.
+    def isBadValDef(t: Tree): Boolean =
+      t match {
+        case (x @ (toolbox.u.ValDef(_, _, _, _) | toolbox.u.DefDef(_, _, _, _, _, _))) =>
+          val content = modifiedContent.substring(x.pos.start, x.pos.end)
+          val prettyPrint = x.toString
```
Comment (eed3si9n) [inline main/src/main/scala/sbt/internals/parser/SbtParser.scala]: @xuwei-k said isn't there a better API than using toString https://twitter.com/xuwei_k/status/520639740274417664. We could check that the last parameter in `ValDef` and `DefDef` are not `EmptyTree` and get the same effect as checking for `=`.

**Comment by eed3si9n** [comment]
Travis is on a roll today.

```
Running run / concurrent
[info] Getting org.scala-sbt sbt 0.13.7-SNAPSHOT ...
[info] :: retrieving :: org.scala-sbt#boot-app
[info]  confs: [default]
[info]  51 artifacts copied, 0 already retrieved (15632kB/93ms)
[info] Getting Scala 2.10.4 (for sbt)...
[info] :: retrieving :: org.scala-sbt#boot-scala
[info]  confs: [default]
[info]  5 artifacts copied, 0 already retrieved (24459kB/67ms)
[info] [error] [/tmp/sbt_1bdda655/concurrent/build.sbt]:35: Pattern matching in val statements is not supported
[info] Project loading failed: (r)etry, (q)uit, (l)ast, or (i)gnore? 
[error] x run / concurrent
```

Looking at https://github.com/sbt/sbt/blob/ddf8a173b0a0bdb8f3f5f129701ec9eba429a5c1/sbt/src/sbt-test/run/concurrent/build.sbt, I'm guessing that it's tripping on:

``` scala
def waitFor(f: File) {
    if(!f.exists) {
        Thread.sleep(300)
        waitFor(f)
    }
}
```

This in fact does _not_ have `=`, but likely the pretty print of Tree would desugar this to

``` scala
def waitFor(...): Unit = { .... }
```

File: main/src/main/scala/sbt/internals/parser/SbtParser.scala [hunk 2]
```diff
@@ -92,6 +92,21 @@ private[sbt] case class SbtParser(file: File, lines: Seq[String]) extends Parsed
         Seq(t)
     }
 
+    // Check No val (a,b) = foo *or* val a,b = foo as these are problematic to range positions and the WHOLE architecture.
+    def isBadValDef(t: Tree): Boolean =
+      t match {
+        case (x @ (toolbox.u.ValDef(_, _, _, _) | toolbox.u.DefDef(_, _, _, _, _, _))) =>
```
Comment (eed3si9n) [inline main/src/main/scala/sbt/internals/parser/SbtParser.scala]: Given the procedure syntax (`def foo {}`), we should just match on `ValDef`, and not `DefDef`.

**Comment by jsuereth** [comment]
@eed3si9n - I just rebased.  Any comments on this pr?

**Comment by eed3si9n** [comment]
LGTM

**Comment by eed3si9n** [comment]
Manually merged this.

**Pull request closed without merging**